retrieving/retrieving-and-querying-mongodb-documents-with-go.md

```diff
@@ -0,0 +1,157 @@
+```
+
+We'll explore marshalling and unmarshalling the response to Go data structures in a later tutorial.
+
+## Reading a Single Document from a Collection
+
+Retrieving all Documents from a collection doesn't always make sense. Instead, sometimes it makes sense to only return a single document. Instead of using the `Find` function, the `FindOne` function can be used.
+
+Take the following example:
+
+```golang
+var podcast bson.D
+err = podcastsCollection.FindOne(ctx, bson.M{}).Decode(&podcast)
+if err != nil {
+    log.Fatal(err)
+}
+fmt.Println(podcast)
+```
+
+In the above example, a `FindOne` is executed without any particular query filter on the data. Rather than returning a Cursor, the single result can be decoded directly into the `bson.D` object. If there is no error, the result will print.
+
+To get an idea of what the result might look like, take the following:
+
+```
+[{_id ObjectID("5d9e0173c1305d2a54eb431a")} {title The Polyglot Developer Podcast} {author Nicolas Raboy}]
+```
+
+Because the results in the `Find` and `FindOne` use `bson.D`, the format of the results will be the same.
+
+## Querying Documents from a Collection with a Filter
+
+In the previous examples of `Find` and `FindOne`, we've seen the `bson.M` filter parameter, even though it wasn't used for filtering. We can make use of the full power of the MongoDB Query Language (MQL) to filter the results of our queries, simply by populating the map.
+
+Let's say that we want to filter our results to only include podcast episodes that are exactly 25 minutes. We could do something like the following:
+
+```golang
+cursor, err := episodesCollection.Find(ctx, bson.M{"duration": 25})
+if err != nil {
+    log.Fatal(err)
+}
+for cursor.Next(ctx) {
+    var episode bson.D
+    cursor.Decode(&episode)
+    fmt.Println(episode)
+}
+cursor.Close(ctx)
+```
+
+Just to reiterate, the full scope of the MongoDB Query Language can be used in our filters, something we'll see in the next example.
+
+## Sorting Documents in a Query
+
+Sorting is a fundamental part of working with data. Rather than sorting the data within the Go application after the query executes, we can let MongoDB do the heavy lifting.
+
+For sorting, we can leverage the `FindOptions` part of the MongoDB Go Driver. `FindOptions` offers more than just sorting, but it is beyond the scope of this getting started example.
+
+Let's say we want to query for all podcast episodes that are longer than 24 minutes, but we want to list them in descending order based on the duration. We could craft a query that looks like the following, while leveraging the `FindOptions` part of the driver:
+
+```golang
+opts := options.Find()
+opts.SetSort(bson.D{{"duration", -1}})
+cursor, err = episodesCollection.Find(ctx, bson.D{{"duration", bson.D{{"$gt", 24}}}}, opts)
+if err != nil {
+    log.Fatal(err)
+}
+for cursor.Next(ctx) {
```

Comment:
See above comments about:
1. Using `cursor.All`
2. Handling the error from `Decode`
3. Calling `Close` in a defer.
4. Using `bson.M` instead of `bson.D`.